### PR TITLE
feat(hotkey): add ringbuf in dsn/utility

### PR DESCRIPTION
**Pull request by Smityz**

### CODE REVIEW

File: include/dsn/utility/ringbuf.h [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+5 7 3
+6 7 2
+7 7 1
+
+*/
+
+namespace dsn {
+namespace utils {
+
+template <typename Type, int MaxElements>
```
Comment (acelyc111) [inline include/dsn/utility/ringbuf.h:44]: Does this class implement yourself?

File: include/dsn/utility/ringbuf.h [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+class ringbuf
```
Comment (acelyc111) [inline include/dsn/utility/ringbuf.h:45]: Could you add some unit tests for it? And I think some benchmark is needed.

File: include/dsn/utility/ringbuf.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+{
```
Comment (acelyc111) [inline include/dsn/utility/ringbuf.h:46]: This class is not thread-safe, it's necessary to add some comment.
And you want to use it for read/write operation hotkey detection, right? Is there any problem if you use it in multi-threads scenario?

File: include/dsn/utility/ringbuf.h [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        return _buf[index];
+    }
+
+    Type front() const { return _buf[_head]; }
+
+    bool empty() const
+    {
+        if (_size != 0)
+            return false;
+        return true;
```
Comment (acelyc111) [inline include/dsn/utility/ringbuf.h:92]:
```suggestion
        return _size == 0;
```

File: include/dsn/utility/ringbuf.h [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+namespace dsn {
+namespace utils {
+
+template <typename Type, int MaxElements>
+class ringbuf
+{
+public:
+    ringbuf() : _size(0), _head(0), _tail(0) {}
+
+    void push(const Type &obj)
```
Comment (acelyc111) [inline include/dsn/utility/ringbuf.h:50]: Is a version of `void push(Type &&obj)` needed?

File: include/dsn/utility/ringbuf.h [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    {
+        _size++;
+        _size = std::min(_size, MaxElements);
+        if (_tail == _head && _size == MaxElements) {
+            _head++;
+            if (_head >= MaxElements) {
+                _head %= MaxElements;
+            }
```
Comment (acelyc111) [inline include/dsn/utility/ringbuf.h:58]: Could you encapsulate these code to a function?

File: include/dsn/utility/ringbuf.h [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+class ringbuf
+{
+public:
+    ringbuf() : _size(0), _head(0), _tail(0) {}
+
+    void push(const Type &obj)
+    {
+        _size++;
+        _size = std::min(_size, MaxElements);
+        if (_tail == _head && _size == MaxElements) {
```
Comment (acelyc111) [inline include/dsn/utility/ringbuf.h:54]: When `_size == MaxElements`, `_tail == _head ` is definitely true, right?

File: include/dsn/utility/ringbuf.h [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            }
+        }
+        _buf[_tail] = obj;
+        _tail++;
+        if (_tail >= MaxElements) {
+            _tail %= MaxElements;
+        }
+    }
+
+    void pop()
```
Comment (acelyc111) [inline include/dsn/utility/ringbuf.h:67]: You should protect when this `ringbuf` is empty.

File: include/dsn/utility/ringbuf.h [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        _head++;
+        _size--;
+        if (_head >= MaxElements) {
+            _head %= MaxElements;
+        }
+        return;
+    }
+
+    Type back() const
+    {
```
Comment (acelyc111) [inline include/dsn/utility/ringbuf.h:78]: What happen when it's empty?

File: include/dsn/utility/ringbuf.h [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        int index = _tail - 1;
+        if (index < 0) {
+            index += MaxElements;
+        }
+        return _buf[index];
+    }
+
+    Type front() const { return _buf[_head]; }
```
Comment (acelyc111) [inline include/dsn/utility/ringbuf.h:86]: What happen when it's empty?

File: include/dsn/utility/ringbuf.h (hunk 1 above)
Comment (Smityz) [reply include/dsn/utility/ringbuf.h:44]: Yes

File: include/dsn/utility/ringbuf.h (hunk 3 above)
Comment (Smityz) [reply include/dsn/utility/ringbuf.h:46]: It's just like `std::queue`, we should add a mutex to keep safety in a multi-threads scenario.

File: include/dsn/utility/ringbuf.h (hunk 7 above)
Comment (Smityz) [reply include/dsn/utility/ringbuf.h:54]: Yes

**Pull request closed without merging**